### PR TITLE
#182201715 users registration with JWT

### PR DESCRIPTION
**Pull request by uwambajeddy**
#### What does this PR do?
This is a user registration PR 

#### Description of Task to be completed?
- creating registration endpoint
- authentication controller

#### How should this be manually tested?
After running the app with `npm run start:dev` use this endpoint `/api/v1/user/signup` on a `POST` request to register a user.
 `GET` `/api/v1/user/` to view inserted user

-before you do anything make sure you have a testing database in your postgres. open a `.env` file and define `TEST_DATABASE=` add your database name and `NODE_ENV=test`
 Run `npm test`

#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/182201715

#### Screenshots (if appropriate)
Screenshot (236)

### CODE REVIEW

**Review by didasmbalanya — CHANGES_REQUESTED** [review]
Missing tests

File: src/controllers/authentication.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    let currentUser = await User.findOne({ where: { email: req.body.email } });
+    if (!currentUser)
+        return next(new AppError('Wrong email or password!', 401));
+
+    const hashedPassword = await compare(req.body.password, currentUser.password);
+
+    if (!hashedPassword)
+        return next(new AppError('Wrong email or password!', 401));
+
+    createSendToken(currentUser, 201, res);
```
Comment (joshwambere) [inline src/controllers/authentication.js]: status should be 200

**Comment by Patricknshimiyimana** [comment]
Hey man, This branch has conflicts that must be resolved

**Comment by uwambajeddy** [comment]
> 

yes, we are merging PRs, when we get on this one I will

**Review by alaintresor — DISMISSED** [review]

**Review by Placide123 — DISMISSED** [review]

File: src/controllers/authentication.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            new AppError('You are not logged in! please login to get access', 401)
+        );
+    }
+    const decoded = await promisify(verify)(token, process.env.JWT_SECRET);
+
+    const currentUser = await User.findByPk(decoded.id);
+
+    if (!currentUser) {
+        return next(
+            new AppError('The user belonging to this token does no longer exist', 401)
```
Comment (joshwambere) [inline src/controllers/authentication.js]: may be a user friendly message, would be great

File: src/helpers/validation_schema.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        .alphanum()
+        .required(),
+
+    lastName: Joi.string()
+        .alphanum()
+        .required(),
+
+    birthDate: Joi.date(),
+
+    phoneNumber: Joi.number(),
```
Comment (joshwambere) [inline src/helpers/validation_schema.js]: the size of the number is crucial here

**Review by joshwambere — CHANGES_REQUESTED** [review]

**Review by joshwambere — APPROVED** [review]

**Pull request merged**